### PR TITLE
KFSPTS-6323 Log exceptions in route step and don't fail job; negate amount on accounting line when object code basic accounting category is AS or EX

### PR DESCRIPTION
**Pull request by db483**

### CODE REVIEW

**Comment by db483** [comment]
Nancy, 

Cathy added a new requirement to this jira: if object code basic category is AS or EX then the amount in the account like should be negated. I reverted the original commit and added coding for this new requirement as well.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AdvanceDepositServiceImpl.java [hunk 1]
```diff
@@ -334,14 +361,32 @@ protected void setupSourceAccountingLine(AchIncomeTransaction transaction, Advan
         sourceAccountingLine.setFinancialObjectCode(objectCode);
         setSourceAccountingLineAccountNumber(account, sourceAccountingLine);
         sourceAccountingLine.setFinancialDocumentLineDescription(transaction.getPayerName());
-        sourceAccountingLine.setAmount(transaction.getTransactionAmount());
+        setSourceAccountingLineAmount(transaction, sourceAccountingLine, chart, objectCode);
```
Comment (db483) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/AdvanceDepositServiceImpl.java]: new code

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AdvanceDepositServiceImpl.java [hunk 2]
```diff
@@ -334,14 +361,32 @@ protected void setupSourceAccountingLine(AchIncomeTransaction transaction, Advan
...
         sourceAccountingLine.setOverrideCode(AccountingLineOverride.CODE.NONE);
         sourceAccountingLine.setPostingYear(advanceDepositDocument.getPostingYear());
         sourceAccountingLine.setDocumentNumber(advanceDepositDocument.getDocumentNumber());
         List<SourceAccountingLine> sourceLines = new ArrayList<>();
         sourceLines.add(sourceAccountingLine);
         advanceDepositDocument.setSourceAccountingLines(sourceLines);
     }
+
```
Comment (db483) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/AdvanceDepositServiceImpl.java]: new code

File: src/main/java/edu/cornell/kfs/sys/CUKFSConstants.java [hunk 3]
```diff
@@ -131,4 +131,9 @@
 
     public static final Set<String> OBJECTS_WITH_IMMEDIATE_BO_LEVEL_NOTE_UPDATE = Collections.singleton("org.kuali.kfs.coa.businessobject.Account");
     
+    public static class BasicAccountingCategory{
```
Comment (db483) [inline src/main/java/edu/cornell/kfs/sys/CUKFSConstants.java]: new code

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AdvanceDepositServiceImpl.java [hunk 4]
```diff
@@ -186,16 +198,31 @@ public boolean routeAdvanceDepositDocuments() {
...
+            }
+        } else {
+            LOG.error(e.getStackTrace());
+        }
     }
+      
 
-    /**
+	/**
      * Returns a list of all initiated but not yet routed advance deposit documents, using the WorkflowDocumentService.
```
Comment (nkimble) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/AdvanceDepositServiceImpl.java]: Small formatting issue. Looks like tabs versus spaces. Will not hold up PR if there are no other issues that need fixing.

Comment (nkimble) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/AdvanceDepositServiceImpl.java]: retest this please

**Comment by nkimble** [comment]
retest this please

**Comment by db483** [comment]
Looks like the unit tests need to be changed after this code change because it is using the bussines object service.

**Comment by nkimble** [comment]
Five unit tests are failing. Attempting a re-run.

**Comment by db483** [comment]
Nancy,

I think I have fixed the issue with the 2 unit tests related to this change. I think the others were dues to sciquest being down and should be fine now. May need to re-run the unit tests again.

Ideally a new unit test should be added to deal with the new test case for this code change but did not get a chance to add that.

Will defer to you and Sandy @se12 to decide wether this can go in as is.

**Comment by nkimble** [comment]
retest this please

**Comment by se12** [comment]
I believe since this is new funtionality, with new code that a new unit test should be created for this scenario.  Thank you.

**Comment by jhulslander** [comment]
with some minor tweeks to setSourceAccountingLineAmount it should be fairly easy to micro-test it  (move extract getting the object and refreshing it to functions, and then it can be easily mocked up)

**Comment by db483** [comment]
Will add the unit test, just did not have time to do it before leaving for the dentist.

**Comment by db483** [comment]
Jay, thank you for the suggestion.

I have added unit tests for the new setSourceAccountingLineAmount method.

**Comment by nkimble** [comment]
Unit tests  added and passed. Merging

**Pull request merged**